usr/src/uts/common/fs/zfs/sys/zap.h

```diff
@@ -87,22 +88,17 @@ extern "C" {
 
 /*
  * Specifies matching criteria for ZAP lookups.
+ * MT_EXACT - exact match, bypasses normalization to do strcmp()
+ * MT_FIRST - Normalizes the lookup - compose/decompose and then
+ *            match the normalized string.
+ * MT_CASE - Used to keep track of whether the lookup request is
```

Comment:
I'm OK with a name change, but it's actually SENSITIVE, not INSENSITIVE.  Sorry for the length, but here's why:

The root of all this is obviously that only the match type of FIRST or EXACT (or the BEST I removed) was kept track of down the stack calls to the normalization and comparison code (and that is what resulted in the flip-flopping in the past of folks fixing case and breaking normalization.)  Thus I had to wedge additional info (CASE) into the match type (mt) as there are 2 cases on a normalized lookup where case matters - I'll refer you to zfs_dropname() (same logic is in zfs_dirent_lock() ) and the header comment above the zfs_dropname function.  Those cases are commented on lines 795 and 798 of zfs_dir.c.  These:

795  * CI  norm | z_norm = upper|formD | mt = MT_FIRST | MT_CASE  (ZCIEXACT)
798  * CM  norm | z_norm = upper|formD | mt = MT_FIRST | MT_CASE  (unix = !ZCILOOK)

Line 795 comments a situation that (probably?) never occurs:  a case insensitive & normalized file system for which the ZCIEXACT flag is passed in?  That implies that we have to normalize (MT_FIRST lookup) but becasue of ZCIEXACT case matters (so we or in MT_CASE.)

Line 798 is more common:  A mixed case & normalized fs.  ZCILOOK is a flag that comes from the Windows (SMB) entry into this lookup path, Unix never sets that flag.  So, if it's a lookup on this fs then case matters (so we or in MT_CASE.)

Now the key use for MT_CASE is down in zap_name_alloc(), and that was the real crux of this problem.  That U8_TEXTPREP_TOUPPER flag gets set on the fs on creation (if I remember correctly?) and it was the normalize and subsequent calls that needed to _preserve_ case on these lookups by removing U8_TEXTPREP_TOUPPER.  MT_CASE is the flag I use to indicate this.

Sorry for the length, but I wanted to make sure it was clear why MT_CASE is sensitive.  That is, a case sensitive lookup on a normalized fs.

So...you OK with changing it to MT_CASE_SENSITIVE ?